book/src/libraries/api_lib_guide.md

```diff
@@ -0,0 +1,319 @@
+  myInOutInt : DINT;
+END_VAR
+END_FUNCTION
+```
+
+```c
+int32_t myFunc(int32_t myInt, char* myString, char* myRefStr, int32_t* myInOutInt);
+```
+
+`FUNCTION_BLOCK`:
+
+
+```iecst
+FUNCTION_BLOCK myFb
+VAR_INPUT
+  myInt  : DINT;
+  myString : STRING[255];
+END_VAR
+
+VAR_IN_OUT
+  myInOutInt : DINT;
+END_VAR
+END_FUNCTION_BLOCK
+```
+
+```c
+
+typedef struct {
+  int32_t myInt;
+  char myString[256];
+  int32_t* myInOutInt;
+} myFbStruct
+
+void myFb(myFbStruct* myFbInstance);
+```
+
+
+#### 2.2.2 Private members
+
+A `FUNCTION_BLOCK` often requires local (private) members to hold data across executions. These members have to be declared in the struct.
+As a side effect, these variables are visible to the users.
+
+For example:
+
+```
+FUNCTION_BLOCK Count
+VAR
+  current : DINT;
+END_VAR
+END_FUNCTION_BLOCK
+```
+
+```c
+
+typedef struct {
+  int32_t current;
+} CountStruct;
+
+void Count(CountStruct* countInst) {
+  countInst->current = countInst->current + 1;
+}
+```
+
+
+
+#### 2.2.3 Return values
+
+A `FUNCTION` defines a return value in the signature, while a `FUNCTION_BLOCK` relies on `VAR_OUTPUT` definitions.
+
+The return type for a function can also include complex datatypes, such as strings, arrays and structs.
+Internally, complex return types are treated as reference parameters.
+
+Example:
+
+```iecst
+FUNCTION myFunc : STRING
+  VAR_INPUT
+    x : DINT;
+  END_VAR
+  VAR_IN_OUT
+    y : DINT;
+  END_VAR
+END_FUNCTION
+```
+
+The C interface would look like:
+
+```c
+void myFunc(char* out, int32_t x, int32_t* y);
+```
+
+A `FUNCTION_BLOCK` should use `VAR_OUTPUT` for return values. Avoid using a
+pointer in the `VAR_INPUT` as a return value.
+
+Example:
+
+```iecst
+FUNCTION_BLOCK myFb
+  VAR_INPUT
+    x : DINT;
+  END_VAR
+  VAR_IN_OUT
+    y : DINT;
+  END_VAR
+  VAR_OUTPUT
+    myOut: DINT;
+    myOut2: STRING[255];
+  END_VAR
+END_FUNCTION
+```
+
+The C interface would look like:
+
+```c
+typedef struct {
+  int32_t x;
+  int32_t* y;
+  int32_t myOut;
+  char myOut2[256];
+
+} myFbStruct;
+
+void myFb(myFbStruct* myFbInst);
+```
+
+### 2.2.4 When to use a `FUNCTION` vs. `FUNCTION_BLOCK`
+
+A `FUNCTION` can be well integrated into the API because of its return value which
+can be nested into expressions. They however don't keep data over subssequent
+executions. If you need to store static data use a `FUNCTION_BLOCK` or use
+`VAR_IN_OUT`.
+
+
+> **_NOTE:_** Do not use `PROGRAM`s in your libraries
+> `PROGRAM`s have static instances. These are reserved for applications and should
+> not be used in libraries.
+
+### 2.3 Datatypes
+
+The IEC61131-3 Standard defines several datatypes with their intended uses.
+To stay standard compliant, an API/Library should try and follow these guidelines.
+
+### 2.3.1 Type sizes
+
+Datatypes are generally convertable to `C` equivalent. With the compiler defaulting to 64bit, some sizes were also fixed to 64bit.
+
+Below is a table of primitive types and how they can be used from `C`
```

Comment:
Are TIME / DATE, strings and pointers primitive datatypes?

Reply:
Arguably C doesn't really have a concept of primitive and non-primitive types but rather primary types and derived/user-defined types - pointers and arrays would be classified as derived types.
As for date/time, they are really just i64s in disguise.

Reply:
I would just call it types then, mainly because we mention strings being complex types somewhere in this document but at the same time call it primitive here.